### ISSUE TITLE
Perl: Can't open perl script

### ISSUE THREAD

**Issue by VitalyKrenel** [so-fancy/diff-so-fancy#295]
Hello. After installing the diff-so-fancy package via npm (with --global flag) I encoutered such a problem: `Can't open perl script ""C:\Users\User\AppData\Roaming\npm\\node_modules\diff-so-fancy\diff-so-fancy""` - it's the exact copy of the message showed in console (I'm using cmder). 
__After removing additional quotes in the diff-so-fancy.cmd (located in "AppData/Roaming/npm") the package is working as it supposed to be.__
From
```perl
perl  "%~dp0\node_modules\diff-so-fancy\diff-so-fancy" %* 
```
to
```perl
perl  %~dp0\node_modules\diff-so-fancy\diff-so-fancy %*
```
Have anyone encountered such a problem?

**Comment by scottchiefbaker** [comment]
I'm not sure of all the nuances of the NPM version (it's maintained by a thirdparty). Can you tried the [standalone version](https://raw.githubusercontent.com/so-fancy/diff-so-fancy/master/third_party/build_fatpack/diff-so-fancy) to see if it works any better?

Also, you might want to try `cmd.exe` or `powershell` there was a bug earlier about cmder, but it may have been unrelated.

**Comment by VitalyKrenel** [comment]
@scottchiefbaker Thanks for your reply. The standalone version should work (another question, wouldn't it be conflicting with the already installed NPM version?). 
The problem I meantioned easily can be removed by changing the bat file inside the npm folder. Again, it isn't a big deal, but it might worsen the experience for ones installing via npm. Is there a way to submit a pull request to their repositry?

**Comment by scottchiefbaker** [comment]
@VitalyKrenel the two should not conflict. Just update your git config to point at the standalone instead of the NPM version.

@OJFord can you comment on the NPM stuff? I don't use, or know anything about NPM, so I can't help much.

**Comment by OJFord** [comment]
> @OJFord can you comment on the NPM stuff? I don't use, or know anything about NPM

I don't either I'm afraid, I've used it a bit a while ago but not enough or recently enough for familiarity, and never to provide dsf.

> Can't open perl script ""C:\Users\User\AppData\Roaming\npm\\node_modules\diff-so-fancy\diff-so-fancy""

I'm suspicious of the double `\\` between 'npm' and 'node_modules' there though - I don't use Windows if I can help it, but that doesn't look right to me? That said the equivalent wouldn't be a problem on Unix, so maybe it's okay.

> perl  %~dp0\node_modules\diff-so-fancy\diff-so-fancy %*

That's not going to work if `%~dp0` contains a special character such as a space, or the location of the script changed to include one (which is outside of your control, it's whatever's in this repo) - so that's not really a solution.

--

If I were you I'd search for similar issues in npm/npm or ask there, it sounds like this is a bug in how you installed it rather than a dsf issue.

**Comment by bimlas** [comment]
Same issue here (Windows). It's funny, because it works on ConEmu and plain PowerShell terminal, Cmd.exe too, but does not works in "embed" terminals (IntelliJ IDEA builtin terminal, Hyper, MSysGit's Bash).

```
$ npm install -g diff-so-fancy
$ git config --global core.pager diff-so-fancy
$ git log
Can't open perl script "/node_modules/diff-so-fancy/diff-so-fancy": No such file or directory
```

Note the forward slashes! It seems that MSysGit does not finds the "Unixish" path, thus the issue seems to related to Git in a combination with some kind of terminals (as I said, it works on ConEmu, but does not works in other terminals). And why does Git thinks that the command is a Perl script?

Versions:
Windows 7 x64 (tried on PowerShell and Cmd.exe)
diff-so-fancy@1.2.0

____

Just tested some possible solutions while I wrote these lines, and adding a bang before the `core.pager` setting in _.gitconfig_ (`pager = "!diff-so-fancy"`) makes it work, but there are other problems with this solution: 

```
$ git config --global core.pager "!diff-so-fancy"
$ git log
# Works as expected, but there is no pagination
```

```
$ git config --global core.pager "!diff-so-fancy | less"
$ git log
!diff-so-fancy | less: !diff-so-fancy: command not found
```

```
$ git config --global core.pager "!sh -c 'diff-so-fancy | less'"
$ git log
!sh -c 'diff-so-fancy | less': !sh: command not found
```

So `!diff-so-fancy` works without any additional flags / pipes.

**Comment by bimlas** [comment]
Just tried to extract the pager in to a shell script, same results:

mypager
```
#!/bin/bash

diff-so-fancy | less
```

```
$ git config --global core.pager mypager
$ git log
Can't open perl script "/node_modules/diff-so-fancy/diff-so-fancy": No such file or directory
```

**Comment by bimlas** [comment]
Omg... (-_-') Use `diff-so-fancy.cmd` instead of `diff-so-fancy` and it will work...

**Comment by bimlas** [comment]
Here's a script which works on Windows and *nix too:

```
$ git config --global core.pager mypager
```

mypager
```
#!/bin/bash
# mypager: Let diff-so-fancy work on Windows in Hyper, IDEATerm, etc.
#
# For details, see
# https://github.com/so-fancy/diff-so-fancy/issues/295#issuecomment-392691782

case "$(uname -s)" in
    CYGWIN*|MINGW*) pagerCmd=diff-so-fancy.cmd;;
    *)              pagerCmd=diff-so-fancy
esac

$pagerCmd | less
```

**Comment by bimlas** [comment]
Just found the original issue which is related to npm + MSys Git which is related to the "startup" script:

c:\Users\USER\AppData\Roaming\npm\diff-so-fancy
```sh
#!/bin/sh
basedir=$(dirname "$(echo "$0" | sed -e 's,\\,/,g')")

case `uname` in
    *CYGWIN*) basedir=`cygpath -w "$basedir"`;;
esac

if [ -x "$basedir/perl" ]; then
  "$basedir/perl"  "$basedir/node_modules/diff-so-fancy/diff-so-fancy" "$@"
  ret=$?
else 
  perl  "$basedir/node_modules/diff-so-fancy/diff-so-fancy" "$@"
  ret=$?
fi
exit $ret
```

The issue is about the first two lines:

```sh
#!/bin/sh
basedir=$(dirname "$(echo "$0" | sed -e 's,\\,/,g')")
```

`basedir` is empty because of the shebang line. Tested on MSys Git Bash:

test/tmp.sh
```sh
#!/bin/sh
echo "A) $0"
echo "B) $0" | sed -e 's,\\,/,g'
echo "C) $(echo "$0" | sed -e 's,\\,/,g')"
echo "D) $(dirname "$(echo "$0" | sed -e 's,\\,/,g')")"
```

Output
```
$ test/tmp.sh
A) test/tmp.sh
B) test/tmp.sh
C)
D)
```

Changing `#!/bin/sh` to `#!/bin/dash`
```
$ test/tmp.sh
A) test/tmp.sh
B) test/tmp.sh
C) test/tmp.sh
D) test
```

I'm not sure where to report: Msys Git, Msys, or npm?

**Comment by bimlas** [comment]
BTW an uggly and very temporary solution is to change `/bin/sh` to `/bin/dash` in your `c:\Users\USER\AppData\Roaming\npm\diff-so-fancy` file.

**Comment by scottchiefbaker** [comment]
This sounds like an msys bug to me? That's **really** odd behavior. Are you sure the line endings are correct? It's not like the shebang line is running in to the second line?

**Comment by bimlas** [comment]
I will try soon (currently on Android), but note that `$(...)` is a Bash expression, not `sh`! The funny part is that is  not working with `#!bash`...

**Comment by bimlas** [comment]
![screencast](https://i.imgur.com/WIVutLp.gif)

**Comment by bimlas** [comment]
Just found the issue on npm: https://github.com/npm/npm/issues/18499

**Comment by scottchiefbaker** [comment]
@bimlas good find.

**Comment by bimlas** [comment]
Found the issue on https://github.com/git-for-windows/git/issues/1619

**Comment by bimlas** [comment]
TL;DR: My antivirus (Comodo Internet Security) caused the problem. Turning it off does not solved the issue, but uninstalling did it.

**Comment by scottchiefbaker** [comment]
Good find. I'll close this issue for now then.
